File: server-rs/src/routes/shared.rs
use std::env;

use serde::Deserialize;

use crate::AppState;
use crate::errors::ApiError;
use crate::models::{AgentProfile, TaskWithPayload};

// ── Shared Path / Query / Payload structs ──

#[derive(Debug, Deserialize)]
pub(crate) struct TaskPath {
    #[serde(rename = "task_id")]
    pub task_id: String,
}

#[derive(Debug, Deserialize)]
pub(crate) struct RepoQuery {
    #[serde(rename = "repoId")]
    pub repo_id: String,
}

#[derive(Debug, Deserialize)]
pub(crate) struct TaskQuery {
    #[serde(rename = "taskId")]
    pub task_id: String,
}

#[derive(Debug, Deserialize)]
pub(crate) struct StartRunPayload {
    #[serde(rename = "planId")]
    pub plan_id: Option<String>,
    #[serde(rename = "taskId")]
    pub task_id: Option<String>,
    #[serde(rename = "profileId")]
    pub profile_id: Option<String>,
    #[serde(rename = "branchName")]
    pub branch_name: Option<String>,
}

// ── Utility functions ──

pub(crate) fn is_todo_lane_status(status: &str) -> bool {
    let upper = status.trim().to_uppercase();
    matches!(
        upper.as_str(),
        "TO DO"
            | "TODO"
            | "PLAN_GENERATING"
            | "PLAN_DRAFTED"
            | "PLAN_APPROVED"
            | "PLAN_REVISE_REQUESTED"
            | "FAILED"
            | "CANCELLED"
    ) || status.to_lowercase().contains("to do")
}

pub(crate) fn sanitize_branch_segment(input: &str) -> String {
    input
        .to_lowercase()
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' || c == '/' {
                c
            } else {
                '-'
            }
        })
        .collect::<String>()
        .split('-')
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("-")
        .chars()
        .take(45)
        .collect()
}

pub(crate) fn home_dir() -> std::path::PathBuf {
    env::var("HOME")
        .or_else(|_| env::var("USERPROFILE"))
        .map(std::path::PathBuf::from)
        .unwrap_or_else(|_| std::path::PathBuf::from("/"))
}

/// Look up the agent command for a repo's preferred agent profile.
pub(crate) fn resolve_agent_command(state: &AppState, repo_id: &str) -> Option<String> {
    let pref = state.db.get_repo_agent_preference(repo_id).ok()??;
    let profile = state
        .db
        .get_agent_profile_by_id(&pref.agent_profile_id)
        .ok()??;
    Some(build_agent_command(&profile))
}

/// 3-tier profile resolution: explicit payload > task override > repo default.
pub(crate) fn resolve_agent_profile(
    state: &AppState,
    explicit_profile_id: Option<&str>,
    task: &TaskWithPayload,
) -> Result<AgentProfile, ApiError> {
    if let Some(profile_id) = explicit_profile_id
        .map(str::trim)
        .filter(|v| !v.is_empty())
    {
        return state
            .db
            .get_agent_profile_by_id(profile_id)
            .map_err(ApiError::internal)?
            .ok_or_else(|| ApiError::bad_request("Agent profile not found."));
    }
    if let Some(ref task_profile_id) = task.agent_profile_id {
        return state
            .db
            .get_agent_profile_by_id(task_profile_id)
            .map_err(ApiError::internal)?
            .ok_or_else(|| ApiError::bad_request("Task agent profile not found."));
    }
    let pref = state
        .db
        .get_repo_agent_preference(&task.repo_id)
        .map_err(ApiError::internal)?
        .ok_or_else(|| ApiError::bad_request("Select an AI profile for this repo."))?;
    state
        .db
        .get_agent_profile_by_id(&pref.agent_profile_id)
        .map_err(ApiError::internal)?
        .ok_or_else(|| ApiError::bad_request("Agent profile not found."))
}

pub(crate) fn build_agent_command(profile: &crate::models::AgentProfile) -> String {
    let command = profile.command.trim();
    if command.is_empty() {
        return profile.command.clone();
    }

    let provider = profile.provider.to_lowercase();
    let model = profile.model.trim();
    if model.is_empty() || model.eq_ignore_ascii_case("default") {
        return command.to_string();
    }

    if provider.contains("codex") {
        return format!("{command} -m {model}");
    }
    if provider.contains("claude")
        || provider.contains("gemini")
        || provider.contains("cursor")
    {
        return format!("{command} --model {model}");
    }

    command.to_string()
}

pub(crate) fn enqueue_autostart_if_enabled(
    state: &AppState,
    task: &TaskWithPayload,
    trigger_kind: &str,
) -> Result<(), ApiError> {
    if !task.auto_start {
        return Ok(());
    }
    if !is_todo_lane_status(&task.status) {
        return Ok(());
    }
    state
        .db
        .enqueue_autostart_job(&task.id, trigger_kind)
        .map_err(ApiError::internal)?;
    Ok(())
}

pub(crate) fn plan_store_key(job_id: &str) -> String {
    format!("plan-job:{job_id}")
}

File: server-ts/src/routes/shared.ts
import crypto from 'crypto';
import os from 'os';
import path from 'path';

import type { Db } from '../db/index.js';
import { ApiError } from '../errors.js';
import type { AgentProfile, RepositoryRule, TaskWithPayload } from '../models.js';
import type { AppState } from '../state.js';

// -- Status helpers --

const TODO_LANE_STATUSES = new Set([
  'TO DO',
  'TODO',
  'FAILED',
  'CANCELLED',
]);

export function isTodoLaneStatus(status: string): boolean {
  const upper = status.trim().toUpperCase();
  if (TODO_LANE_STATUSES.has(upper)) {
    return true;
  }
  return status.toLowerCase().includes('to do');
}

// -- Branch sanitization --

export function sanitizeBranchSegment(input: string): string {
  const replaced = input
    .split('')
    .map((c) => {
      if (/[A-Za-z0-9._\-/]/.test(c)) {
        return c;
      }
      return '-';
    })
    .join('');

  return replaced
    .split('-')
    .filter((s) => s.length > 0)
    .join('-')
    .slice(0, 60);
}

// -- Home directory --

export function homeDir(): string {
  return os.homedir() || process.env['HOME'] || process.env['USERPROFILE'] || os.tmpdir();
}

// -- Application data directory --

/** Resolve the per-user data directory for this app, matching the DB location logic. */
export function getAppDataDir(): string {
  const override = process.env['APP_DATA_DIR'];
  if (override) return override;

  const platform = os.platform();
  if (platform === 'darwin') {
    return path.join(homeDir(), 'Library', 'Application Support', 'branching-bad');
  }
  if (platform === 'win32') {
    return path.join(process.env['APPDATA'] || path.join(homeDir(), 'AppData', 'Roaming'), 'branching-bad');
  }
  // Linux / other
  return path.join(process.env['XDG_DATA_HOME'] || path.join(homeDir(), '.local', 'share'), 'branching-bad');
}

/** Stable, human-readable namespace for a repo on disk: `<basename>-<shortHash>`. */
export function repoNamespace(repoPath: string): string {
  const normalized = path.resolve(repoPath);
  const base = path.basename(normalized).replace(/[^A-Za-z0-9._-]/g, '-') || 'repo';
  const hash = crypto.createHash('sha1').update(normalized).digest('hex').slice(0, 8);
  return `${base}-${hash}`;
}

/** Where worktrees for a given repo live, outside the repo itself. */
export function worktreesRootFor(repoPath: string): string {
  return path.join(getAppDataDir(), 'worktrees', repoNamespace(repoPath));
}

// -- Agent command resolution --

export function resolveAgentCommand(state: AppState, repoId: string): string | null {
  const pref = state.db.getRepoAgentPreference(repoId);
  if (!pref) {
    return null;
  }
  const profile = state.db.getAgentProfileById(pref.agent_profile_id);
  if (!profile) {
    return null;
  }
  return buildAgentCommand(profile);
}

// -- 3-tier agent profile resolution --

export function resolveAgentProfile(
  state: AppState,
  explicitProfileId: string | undefined | null,
  task: TaskWithPayload,
): AgentProfile {
  const trimmedExplicit = explicitProfileId?.trim();
  if (trimmedExplicit) {
    const profile = state.db.getAgentProfileById(trimmedExplicit);
    if (!profile) {
      throw ApiError.badRequest('Agent profile not found.');
    }
    return profile;
  }

  if (task.agent_profile_id) {
    const profile = state.db.getAgentProfileById(task.agent_profile_id);
    if (!profile) {
      throw ApiError.badRequest('Task agent profile not found.');
    }
    return profile;
  }

  const pref = state.db.getRepoAgentPreference(task.repo_id);
  if (!pref) {
    throw ApiError.badRequest('Select an AI profile for this repo.');
  }

  const profile = state.db.getAgentProfileById(pref.agent_profile_id);
  if (!profile) {
    throw ApiError.badRequest('Agent profile not found.');
  }
  return profile;
}

// -- Build agent command with model flags --

export function buildAgentCommand(profile: AgentProfile): string {
  const command = profile.command.trim();
  if (!command) {
    return profile.command;
  }

  const provider = profile.provider.toLowerCase();
  const model = profile.model.trim();

  if (!model || model.toLowerCase() === 'default') {
    return command;
  }

  if (provider.includes('codex')) {
    return `${command} -m ${model}`;
  }

  if (provider.includes('claude') || provider.includes('gemini') || provider.includes('cursor')) {
    return `${command} --model ${model}`;
  }

  return command;
}

// -- Autostart queueing --

export function enqueueAutostartIfEnabled(
  state: AppState,
  task: TaskWithPayload,
  triggerKind: string,
): void {
  if (!task.auto_start) {
    return;
  }
  if (!isTodoLaneStatus(task.status)) {
    return;
  }
  state.db.enqueueAutostartJob(task.id, triggerKind);
}

// -- Plan store key --

export function planStoreKey(jobId: string): string {
  return `plan-job:${jobId}`;
}

// -- Persist store outputs to DB --

export function persistStoreOutputs(
  store: import('../msgStore.js').MsgStore,
  db: import('../db/index.js').Db,
  taskId: string,
): () => void {
  const unsubscribe = store.subscribe((msg) => {
    try {
      db.pushTaskOutput(taskId, msg.type, msg.data);
    } catch {
      // Ignore DB write failures for output logs
    }
    if (msg.type === 'finished') {
      unsubscribe();
    }
  });
  return unsubscribe;
}

// -- Rules loading --

export function loadRulesSection(db: Db, repoId: string): string {
  let rules: RepositoryRule[];
  try {
    rules = db.listRulesForPrompt(repoId);
  } catch {
    rules = [];
  }
  return formatRulesPromptSection(rules);
}

export function formatRulesPromptSection(rules: RepositoryRule[]): string {
  if (rules.length === 0) {
    return '';
  }
  const lines = rules.map((r) => `- ${r.content}`);
  return `\n\nRepository Rules (follow these strictly):\n${lines.join('\n')}\n`;
}
